File: java/src/main/java/vsdk/toolkit/common/dataStructures/KDTreeNode.java
package vsdk.toolkit.common.dataStructures;

public class KDTreeNode {
    public KDTreeNode loson;
    public KDTreeNode hison;

    public int mFlags;
    public Object mData;

    public int discriminator() {
        return mFlags & 0xF;
    }

    public void setDiscriminator(int discriminator) {
        mFlags = (mFlags & 0xFFF0) | discriminator;
    }

    public int flags() {
        return mFlags & 0xFFF0;
    }

    public void findMinMaxDepth(int depth, int[] minDepth, int[] maxDepth) {
        if ((loson == null) && (hison == null)) {
            maxDepth[0] = Math.max(maxDepth[0], depth);
            minDepth[0] = Math.min(minDepth[0], depth);
        }
        else {
            if (loson != null) {
                loson.findMinMaxDepth(depth + 1, minDepth, maxDepth);
            }
            if (hison != null) {
                hison.findMinMaxDepth(depth + 1, minDepth, maxDepth);
            }
        }
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/dataStructures/KDTreeNode.ts
export class KDTreeNode {
  public loson: KDTreeNode | null;
  public hison: KDTreeNode | null;

  public mFlags: number;
  public mData: unknown;

  public constructor() {
    this.loson = null;
    this.hison = null;
    this.mFlags = 0;
    this.mData = null;
  }

  public discriminator(): number {
    return this.mFlags & 0xF;
  }

  public setDiscriminator(discriminator: number): void {
    this.mFlags = (this.mFlags & 0xFFF0) | discriminator;
  }

  public flags(): number {
    return this.mFlags & 0xFFF0;
  }

  public findMinMaxDepth(depth: number, minDepth: number[], maxDepth: number[]): void {
    if ((this.loson === null) && (this.hison === null)) {
      maxDepth[0] = globalThis.Math.max(maxDepth[0], depth);
      minDepth[0] = globalThis.Math.min(minDepth[0], depth);
    }
    else {
      if (this.loson !== null) {
        this.loson.findMinMaxDepth(depth + 1, minDepth, maxDepth);
      }
      if (this.hison !== null) {
        this.hison.findMinMaxDepth(depth + 1, minDepth, maxDepth);
      }
    }
  }
}
